### PR TITLE
Lesson 16: Test User Registration Form

### PR DESCRIPTION
**Pull request by olyastrilectka**

### CODE REVIEW

File: cypress/constants/user-data.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  maths: 'Maths',
+  accounting: 'Accounting',
+  arts: 'Arts',
+  biology: 'Biology',
+  computerScience: 'Computer Science',
+  hindi: 'Hindi',
+  commerce: 'Commerce',
+  history: 'History',
+};
+const HOBBIES = {
```
Comment (landriiuk) [inline cypress/constants/user-data.js]: missed space

File: cypress/integration/homework16-female-gender.spec.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+  randomUserData,
+  GENDER,
+  SUBJECT,
+  HOBBIES,
+  STATES,
+  UTTAR_PRADESH_STATE_CITIES,
+} from '../constants/user-data';
+
+describe('Lesson16', () => {
+  describe('Female Gender Flow', () => {
```
Comment (landriiuk) [inline cypress/integration/homework16-female-gender.spec.js]: u can use context for the second describe

File: cypress/integration/homework16-female-gender.spec.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    let registrationFormPage;
+    let submitStudentDataForm;
+
+    before(() => {
+      registrationFormPage = new RegistrationFormPage();
+      registrationFormPage.open();
+
+      submitStudentDataForm = new SubmitStudentDataForm();
+    });
```
Comment (landriiuk) [inline cypress/integration/homework16-female-gender.spec.js]:
```suggestion
    let registrationFormPage = new RegistrationFormPage();
    let submitStudentDataForm = new SubmitStudentDataForm();

    before(() => {
      registrationFormPage.open();
    });
```

File: cypress/integration/homework16-male-gender.spec.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  describe('Male Gender Flow', () => {
+    let registrationFormPage;
+    let submitStudentDataForm;
+
+    before(() => {
+      registrationFormPage = new RegistrationFormPage();
+      registrationFormPage.open();
+
+      submitStudentDataForm = new SubmitStudentDataForm();
+    });
```
Comment (landriiuk) [inline cypress/integration/homework16-male-gender.spec.js]: Same here as at cypress/integration/homework16-female-gender.spec.js

File: cypress/integration/homework16-male-gender.spec.js [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+      registrationFormPage.lastNameInput
+        .type(randomUserData.FAKE_LAST_NAME)
+        .should('have.value', randomUserData.FAKE_LAST_NAME);
+    });
+
+    it('Should fill Email input', () => {
+      registrationFormPage.emailInput
+        .type(randomUserData.FAKE_EMAIL)
+        .should('have.value', randomUserData.FAKE_EMAIL);
+    });
```
Comment (landriiuk) [inline cypress/integration/homework16-male-gender.spec.js:42]: missed empty row

File: cypress/integration/homework16-male-gender.spec.js [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+      registrationFormPage.maleGenderRadioButton.click({ force: true });
+      registrationFormPage.maleGenderRadioButton.should('be.checked');
+    });
+
+    it('Should fill Mobile input', () => {
+      registrationFormPage.mobileInput
+        .type(randomUserData.FAKE_MOBILE)
+        .should('have.value', randomUserData.FAKE_MOBILE);
+    });
+    it('Should fill DatePicker input', () => {
```
Comment (landriiuk) [inline cypress/integration/homework16-male-gender.spec.js:55]: and here

File: cypress/integration/homework16-male-gender.spec.js [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+      registrationFormPage.selectFileInput.selectFile(
+        `${fixtureFolderPath}/photo.png`,
+      );
+    });
+
+    it('Should fill Current Address field', () => {
+      registrationFormPage.currentAddress
+        .type(randomUserData.FAKE_CURRENT_ADDRESS)
+        .should('have.value', randomUserData.FAKE_CURRENT_ADDRESS);
+    });
```
Comment (landriiuk) [inline cypress/integration/homework16-male-gender.spec.js:93]: and here

File: cypress/integration/homework16-male-gender.spec.js [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    it('Should select value from State dropdown', () => {
+      registrationFormPage.stateDropdown.click();
+      registrationFormPage.dropdownMenu.contains(STATES.ncr).click();
+    });
+    it('Should select value from City dropdown', () => {
```
Comment (landriiuk) [inline cypress/integration/homework16-male-gender.spec.js:100]: and here

File: cypress/integration/homework16-other-gender.spec.js [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+  describe('Other Gender Flow', () => {
+    let registrationFormPage;
+    let submitStudentDataForm;
+
+    before(() => {
+      registrationFormPage = new RegistrationFormPage();
+      registrationFormPage.open();
+
+      submitStudentDataForm = new SubmitStudentDataForm();
+    });
```
Comment (landriiuk) [inline cypress/integration/homework16-other-gender.spec.js]: About describe and about variables, added comments in cypress/integration/homework16-female-gender.spec.js

File: cypress/integration/homework16-female-gender.spec.js [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+    it('Should check Hobbies checkbox', () => {
+      registrationFormPage.musicHobbyCheckbox.click({ force: true });
+      registrationFormPage.hobbyLabels.contains(HOBBIES.music).should('exist');
+    });
+
+    it('Should download the picture', () => {
+      const fixtureFolderPath = Cypress.config('fixturesFolder');
+      registrationFormPage.selectFileInput.selectFile(
+        `${fixtureFolderPath}/sunset.jpg`,
```
Comment (landriiuk) [inline cypress/integration/homework16-female-gender.spec.js]: U can create it as a method, after that using in your tests.

File: cypress/page-objects/student-registration-form.js [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+  subjectsDropdownSelector = '#subjectsContainer';
+
+  dropdownSubjectMenuSelector = 'div[id*="option"]';
+
+  subjectLabelsSelector = '.subjects-auto-complete__multi-value__label';
+
+  selectFileInputSelector = '#uploadPicture';
+
+  get submitBtn() {
+    return cy.get(this.submitButtonSelector);
```
Comment (landriiuk) [inline cypress/page-objects/student-registration-form.js]: To my mind better write ### 
```suggestion
    return cy.get('#submit');
```
and delete part of code at the top

File: cypress/page-objects/submit-student-data-modal.js [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    return cy.get(this.submittedStudentAddressValueSelector);
+  }
+
+  get studentStateAndCityValue() {
+    return cy.get(this.submittedStudentStateAndCityValueSelector);
+  }
+}
+
+// eslint-disable-next-line import/prefer-default-export
+export { SubmitStudentDataForm };
```
Comment (landriiuk) [inline cypress/page-objects/submit-student-data-modal.js:70]: here also about selectors.

**Review by landriiuk — APPROVED** [review]

**Comment by landriiuk** [comment]
To my mind, it is unnecessary to write many tests on a small action.
I'd rather create multiple tests.

A few things to improve. you have a lot of copied code. Create a folder of models and create all reusable methods.(it will be like a part of POM)

Your code working well 
Score 9/10

Check comments and merge this branch

File: cypress/constants/user-data.js (hunk 1 above)
Comment (olyastrilectka) [reply cypress/constants/user-data.js]: thanks

File: cypress/integration/homework16-female-gender.spec.js (hunk 2 above)
Comment (olyastrilectka) [reply cypress/integration/homework16-female-gender.spec.js]: good idea

File: cypress/integration/homework16-female-gender.spec.js (hunk 3 above)
Comment (olyastrilectka) [reply cypress/integration/homework16-female-gender.spec.js]: thanks, I have changed

File: cypress/integration/homework16-male-gender.spec.js (hunk 4 above)
Comment (olyastrilectka) [reply cypress/integration/homework16-male-gender.spec.js]: thanks, good idea, I have changed my code

File: cypress/integration/homework16-male-gender.spec.js (hunk 5 above)
Comment (olyastrilectka) [reply cypress/integration/homework16-male-gender.spec.js:42]: thanks

File: cypress/integration/homework16-male-gender.spec.js (hunk 8 above)
Comment (olyastrilectka) [reply cypress/integration/homework16-male-gender.spec.js:100]: thx

File: cypress/integration/homework16-other-gender.spec.js (hunk 9 above)
Comment (olyastrilectka) [reply cypress/integration/homework16-other-gender.spec.js]: thx, I have changed

File: cypress/integration/homework16-female-gender.spec.js (hunk 10 above)
Comment (olyastrilectka) [reply cypress/integration/homework16-female-gender.spec.js]: changed

File: cypress/page-objects/student-registration-form.js (hunk 11 above)
Comment (olyastrilectka) [reply cypress/page-objects/student-registration-form.js]: did it

File: cypress/page-objects/submit-student-data-modal.js (hunk 12 above)
Comment (olyastrilectka) [reply cypress/page-objects/submit-student-data-modal.js:70]: thanks, good idea, I have changed my code

**Pull request merged**